src/Passes/docs/base-profile-transformations.md

```diff
@@ -0,0 +1,233 @@
+  call void @__quantum__rt__array_update_alias_count(%Array* %qs, i32 -1)
+  call void @__quantum__rt__qubit_release_array(%Array* %qs)
+  ret void
+}
+```
+
+After applying the our [demo profile transformation](https://github.com/troelsfr/qsharp-compiler/tree/feature/profile-pass/src/Passes/examples/QubitAllocationAnalysis), the QIR is reduced to:
+
+```
+define void @Feasibility__QubitMapping__Interop() local_unnamed_addr #0 {
+entry:
+  call void @__quantum__qis__x__body(%Qubit* null)
+  call void @__quantum__qis__x__body(%Qubit* nonnull inttoptr (i64 1 to %Qubit*))
+  call void @__quantum__qis__x__body(%Qubit* nonnull inttoptr (i64 2 to %Qubit*))
+  ret void
+}
+```
+
+We note that we successfully have eliminated loops, arithmetic operations, dynamic qubit allocation and alias counting - all operations which are not supported by the base profile.
+
+## Goal
+
+We envision the tool to work as a standalone commandline tool which can either validate or generate a QIR in accordance with a given profile. To validate, one would run:
+
+```language
+qat -p profile.yaml --validate unvalidated-qir.ll
+```
+
+In a similar fashion, generation is performed by adding `--generate` to the commandline:
+
+```language
+qat -p profile.yaml --generate qir.ll > qir-profile.ll
+```
+
+Default behaviour of the tool is that it always validates the generated profile. This behaviour can be disabled by
+
+```language
+qat -p profile.yaml --generate --no-validate qir.ll > qir-profile.ll
+```
+
+# Profile Specification
+
+Every profile is specified through a YAML file which defines an object at the top-level. This object must contain the fields `name` and `displayName`:
+
+```yaml
+name: profile-name
+displayName: Profile Name
+# ...
+```
+
+Additionally, top level also contains the fields `version` and `mode`. The version refers to the QIR version which forms the basis for the specification and `mode` explains how the profile is defined.
+The final two top level fields are lists named `specification` and `generation`. These contains the specification and generation producedure, respectively.
+
+## Specification
+
+The default `mode` of specification is by `feature` which means that specification describes the feature set available. Alternatively, one can specify a profile by `limitation`. As an example profile for
+
+```yaml
+name: profile-name
+displayName: Profile Name
+version: 1.0
+mode: feature
+specification:
+  functions:
+    - __quantum__qis__toffoli__body
+    - __quantum__qis__cnot__body
+    - __quantum__qis__cz__body
+    - __quantum__qis__h__body
+    - __quantum__qis__mz__body
+    - __quantum__qis__reset__body
+    - __quantum__qis__rx__body
+    - __quantum__qis__ry__body
+    - __quantum__qis__rz__body
+    - __quantum__qis__s__body
+    - __quantum__qis__s_adj
+    - __quantum__qis__t__body
+    - __quantum__qis__t__adj
+    - __quantum__qis__x__body
+    - __quantum__qis__y__body
+    - __quantum__qis__z__body
+  instructions:
```

Comment:
What about specifying which classical values/types are supported? Within the base profile, int and double values are supported, but not anything else. How would one specify e.g. the availability of fixed point numbers with a certain precision (only)?

Reply:
I think this is an interesting question that possibly requires a wider discussion.